File: handlerclaw/lib/features/prayer-logs/data/dto/prayer_log_dto.dart
class PrayerLogDto {
  final String id;
  final String userId;
  final DateTime date;
  final String prayer;
  final String category;
  final bool performed;
  final DateTime performedAt;
  final bool isQadha;
  final String method;
  final String place;
  final String? notes;

  PrayerLogDto({
    required this.id,
    required this.userId,
    required this.date,
    required this.prayer,
    required this.category,
    required this.performed,
    required this.performedAt,
    required this.isQadha,
    required this.method,
    required this.place,
    this.notes,
  });

  factory PrayerLogDto.fromJson(Map<String, dynamic> json) {
    return PrayerLogDto(
      id: json['id'] ?? '',
      userId: json['user_id'] ?? '',
      date: DateTime.parse(json['date'] ?? DateTime.now().toIso8601String()),
      prayer: json['prayer'] ?? '',
      category: json['category'] ?? '',
      performed: json['performed'] ?? false,
      performedAt: DateTime.parse(
        json['performed_at'] ?? DateTime.now().toIso8601String(),
      ),
      isQadha: json['is_qadha'] ?? false,
      method: json['method'] ?? 'SENDIRI',
      place: json['place'] ?? 'RUMAH',
      notes: json['notes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'date': date.toIso8601String(),
      'prayer': prayer,
      'category': category,
      'performed': performed,
      'performed_at': performedAt.toIso8601String(),
      'is_qadha': isQadha,
      'method': method,
      'place': place,
      'notes': notes,
    };
  }
}

File: api/src/modules/prayer-log/application/dtos/prayer-log.dto.ts
import type {
  PrayerCategory,
  PrayerMethod,
  PrayerPlace,
  PrayerType,
} from "../../../../lib/generated/prisma/client.js";
import type { IPrayerLog } from "../../domain/entities/prayer-log.entity.js";
import type { PaginationMetaDto } from "../../../../lib/types/pagination-meta-dto.js";

/**
 * Input Data Contracts
 */
export type CreatePrayerLogRequest = {
  prayer: PrayerType;
  local_date?: Date;
  performed_at?: Date;
  method?: PrayerMethod;
  place?: PrayerPlace;
  is_qadha: boolean;
  notes?: string;
};

export type CreatePrayerLogData = {
  userId: string;
  date: Date;
  prayer: PrayerType;
  performed: boolean;
  performedAt?: Date | null;
  category: PrayerCategory;
  method?: PrayerMethod | null;
  place?: PrayerPlace;
  isQadha?: boolean;
  notes?: string | null;
};

export type UpdatePrayerLogData = {
  performed?: boolean;
  performedAt?: Date | null;
  category?: PrayerCategory;
  method?: PrayerMethod | null;
  place?: PrayerPlace;
  isQadha?: boolean;
  notes?: string | null;
};

export type GetPrayerLogsQuery = {
  page: number;
  per_page: number;
  date_type:
    | "today"
    | "this_week"
    | "this_month"
    | "this_year"
    | "7_days"
    | "30_days"
    | "1_year"
    | "all"
    | "custom";
  start?: string;
  end?: string;
};

export type GetPrayerLogsSummaryQuery = {
  date_type:
    | "today"
    | "this_week"
    | "this_month"
    | "this_year"
    | "7_days"
    | "30_days"
    | "1_year"
    | "all"
    | "custom";
  start?: string;
  end?: string;
};

/**
 * Filter Contracts
 */
export type DateRangeFilter = {
  gte?: Date | string;
  lte?: Date | string;
  gt?: Date | string;
  lt?: Date | string;
};

export type PrayerLogRepositoryFilter = {
  date?: DateRangeFilter;
  performed?: boolean;
  prayer?: PrayerType;
};

export type ByPrayer = Record<
  PrayerType,
  {
    performed: number;
    qadha: number;
    jamaah: number;
  }
>;

export type PrayerSummaryEntry = {
  label: string;
  total_wajib_performed: number;
  total_sunnah_performed: number;
  total_performed: number;
  total_qadha: number;
  by_prayer: ByPrayer;
};

/**
 * Response Contracts
 */
export type CreatePrayerLogResponse = {
  prayer_log: IPrayerLog;
};

export type GetPrayerLogsResponse = {
  filter: {
    date_type: string;
    start?: string | undefined;
    end?: string | undefined;
    filtered?: {
      date_start?: string | undefined;
      date_end?: string | undefined;
    };
  };
  prayer_logs: IPrayerLog[];
  meta: PaginationMetaDto;
};

export type GetPrayerLogsSummaryResponse = {
  filter: {
    date_type: string;
    start?: string | undefined;
    end?: string | undefined;
    filtered?: {
      date_start?: string | undefined;
      date_end?: string | undefined;
    };
  };
  summary: PrayerSummaryEntry[];
  count: Record<string, number>;
  performances: Record<string, { count: number; percentage?: number }>;
  grand_total: {
    total_wajib_performed: number;
    total_sunnah_performed: number;
    total_performed: number;
    total_qadha: number;
  };
};
